### ISSUE TITLE
Some character not being correctly typed

### ISSUE THREAD

**Issue by victor141516** [nut-tree/nut.js#419]
**Version**

2.1.1

**Short overview**

**Issue occurs on**
- [ ] Virtual machine
- [x] Docker container
- [ ] Dev/Host system

**Detailed error description**

I'm trying to type some characters like `@` or `#` but the corresponding number is being typed instead (2 and 3 in this case)

I suppose the error is caused because the modifier keys are not being pressed

**Steps to reproduce error**

I'm using a Dockerfile you can find here: https://github.com/victor141516/bipbop/blob/master/Dockerfile

Instructions to build and run it can be found here: https://github.com/victor141516/bipbop#how-to-use

That image exposes an HTTP API you can use to type using this request:
```bash
curl -s -XPOST -H 'content-type: application/json' -d '[["##@@"]]' http://localhost:3000/api/v1/type
```

This API is just calling `keyboard.type('##@@')`: https://github.com/victor141516/bipbop/blob/master/pilot/src/services/browser/index.ts#L128-L130

Then you can check the browser on https://localhost:6901/ (all this is explained in the readme)

My image is based on [kasmweb/chrome](https://hub.docker.com/r/kasmweb/chrome) so you can use that and install node on it instead

---

EDIT: I forgot to mention I can send Shift + 2 (using key codes) to type `@`. For that I'm using:
```bash
curl -s -XPOST -H 'content-type: application/json' -d '[[7, 37]]' http://localhost:3000/api/v1/type
```

That will execute `keyboard.type([7, 37]);`

**Comment by victor141516** [comment]
I tried using robotjs and the result is just the same (@ becomes 2)

So I suppose the problem is related to the OS and not the library. Do you have any idea? Surely you have more experience handling these things 😅

**Comment by s1hofmann** [comment]
Hey @victor141516 👋 

nut.js suffers the same problem as robot.js.
On Linux, when typing strings, the X11 API does not apply any modifiers, thus presses the unmodified key instead.
`2` instead of `@` in case of US ISO layout.

Since this issue is a duplicate of #239, I'll close this one and keep #239 updated

**Comment by victor141516** [comment]
Mmmm but why I can use `[7, 37]` (Left shift + 2) and I get an `@` ?

I was thinking of writting my own translation function (`@` --> `[7,37]`) before submitting this issue. I can work on that and submit a PR, but I now looks like I'm missing something.

**Comment by s1hofmann** [comment]
Hi @victor141516 👋 

Let me supply some more context.

The way the Linux `type` implementation for typing strings works is that for each character in the string it determines the key symbol and from there, the key code.
Now the first problem is that multiple symbols can be mapped to the same code, using modifiers and both keys and modifiers can be remapped as well.
Unfortunately, the API I mentioned earlier does not return required modifier keys, so this would be something nut.js (or more specific, libnut) would have to resolve manually.

The second problem is that if there's no symbol registered for a character, this whole approach fails, which is the main problem with unicode support.

To sum things up: The approach you mentioned works for your case, because you're sending the correct key + modifier for US ANSI layout to type '@'.
But given the amount of customisability, which is often used, it's not suitable for a general solution.

There's a lot to improve regarding Linux keyboard support, but this will require quite a bit of time and additional knowledge to do. At least for me, as I'm not that familiar with XLib programming.

Best regards

Simon

**Comment by victor141516** [comment]
Thank you so much for that clarifier comment 💯
